File: python/packages/hoermoles-ble/src/hoermoles_ble/bundle.py
"""
Transportable credential bundle: the format used to move taught drives
(credentials + the non-secret product metadata) between this Python
implementation and other clients - primarily the web app, which cannot read
`~/.hoermoles/credentials/*.json` off the filesystem.

Deliberately a *separate, versioned wire format* rather than "just ship the
credentials file": a bundle can hold several drives at once, carries the
`devices.json` product metadata along (so the web app knows which menu table
applies without a BLE scan), and has a text form that fits into a QR code or a
URL fragment.

Two encodings, both round-trippable by the TypeScript port in
`spa/packages/hoermoles-ble-js/src/bundle.ts` - keep the two in sync, they
implement one spec:

    HMOLES1:<base64url(utf8 json)>            plaintext
    HMOLES1E:<base64url(binary envelope)>     passphrase-encrypted

The encrypted envelope is
`b"HM1E" || salt(16) || nonce(12) || AES-256-GCM(ciphertext||tag)`, key derived
via PBKDF2-HMAC-SHA256. Encryption is not optional decoration: a root key is a
physical door key, and the text form is meant to travel through QR codes, URL
fragments and chat messages, all of which get logged, cached and shoulder-surfed.
Plaintext export therefore stays available (it is convenient and often fine for
a QR code shown on a local screen) but callers should default to encrypting
anything that leaves the machine.

base64url is used without padding so the text form is safe in a URL fragment.

`serial_no` is carried as a *string*, not a JSON number, on purpose: Hoermann
serial numbers are uint64 (the live test device's is 302626026414510307), which
exceeds JavaScript's Number.MAX_SAFE_INTEGER. A JSON number would silently lose
its last digits when the web app parses it - caught exactly this way by the
shared test vectors.
"""

from __future__ import annotations

import base64
import json
import secrets
from dataclasses import dataclass
from typing import Any

from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC

from .credentials import Credentials
from .devices import DeviceInfo

BUNDLE_FORMAT = "hoermoles-credentials"
BUNDLE_VERSION = 1

PREFIX_PLAIN = "HMOLES1:"
PREFIX_ENCRYPTED = "HMOLES1E:"

_ENCRYPTED_MAGIC = b"HM1E"
_SALT_SIZE = 16
_NONCE_SIZE = 12
_KEY_SIZE = 32
# OWASP's 2023 floor for PBKDF2-HMAC-SHA256 is 600k; WebCrypto does the same work in
# the browser in well under a second, so there is no reason to go below it here.
_PBKDF2_ITERATIONS = 600_000


class BundleError(ValueError):
    """Malformed bundle: wrong prefix, bad base64, unknown version, or - for the
    encrypted form - a wrong passphrase (AES-GCM authentication failure)."""


@dataclass
class BundleEntry:
    """One drive in a bundle: the secret part (Credentials) plus, when known,
    the non-secret registry part (DeviceInfo). `device_info` is optional because
    a drive can have been registered without ever being scanned."""

    credentials: Credentials
    device_info: DeviceInfo | None = None

    def to_dict(self) -> dict[str, Any]:
        entry: dict[str, Any] = {
            "device_address": self.credentials.device_address.upper(),
            "root_id": self.credentials.root_id,
            "root_key_hex": self.credentials.root_key.hex(),
            "qr_prefix": self.credentials.qr_prefix,
            "created_unix": self.credentials.created_unix,
        }
        # A user-chosen display name. Top-level (not under the product block), to
        # mirror the TypeScript port's flat layout and so it round-trips even
        # when the product type is unknown. On this side the value lives in
        # DeviceInfo.label, so it is only present once a drive has a DeviceInfo.
        if self.device_info is not None and self.device_info.label:
            entry["label"] = self.device_info.label
        if self.device_info is not None:
            entry["product_class"] = self.device_info.product_class
            entry["product_id"] = self.device_info.product_id
            entry["product_name"] = self.device_info.product_name
            # str(), not int - see the module docstring on uint64 vs JS numbers.
            serial_no = self.device_info.serial_no
            entry["serial_no"] = None if serial_no is None else str(serial_no)
        return entry

    @classmethod
    def from_dict(cls, entry: dict[str, Any]) -> BundleEntry:
        try:
            address = entry["device_address"]
            credentials = Credentials(
                device_address=address,
                root_id=int(entry["root_id"]),
                root_key=bytes.fromhex(entry["root_key_hex"]),
                qr_prefix=entry.get("qr_prefix", ""),
                created_unix=int(entry.get("created_unix", 0)),
            )
        except (KeyError, TypeError, ValueError) as exc:
            raise BundleError(f"Malformed bundle entry: {exc}") from exc

        device_info = None
        if entry.get("product_class") is not None and entry.get("product_id") is not None:
            serial_no = entry.get("serial_no")
            device_info = DeviceInfo(
                device_address=address.upper(),
                product_class=int(entry["product_class"]),
                product_id=int(entry["product_id"]),
                product_name=entry.get("product_name"),
                # Accepts the legacy JSON-number form too, so a bundle written
                # before the string change still imports.
                serial_no=None if serial_no is None else int(serial_no),
                label=entry.get("label"),
            )
        # A label without any product info (theoretically possible from the web
        # app, though a registered drive always has product info) has nowhere to
        # be stored on this side - DeviceInfo requires a product - so it is
        # dropped rather than fabricating a placeholder product type.
        return cls(credentials=credentials, device_info=device_info)


def build_bundle(entries: list[BundleEntry]) -> dict[str, Any]:
    """The bundle as a plain dict - the single canonical JSON shape, shared by
    the file form, the plaintext text form and (after encryption) the encrypted
    text form."""
    return {
        "format": BUNDLE_FORMAT,
        "v": BUNDLE_VERSION,
        "devices": [entry.to_dict() for entry in entries],
    }


def parse_bundle(payload: dict[str, Any]) -> list[BundleEntry]:
    if not isinstance(payload, dict):
        raise BundleError("Bundle must be a JSON object.")
    if payload.get("format") != BUNDLE_FORMAT:
        raise BundleError(f"Not a hoermoles credential bundle (format={payload.get('format')!r}).")
    version = payload.get("v")
    if version != BUNDLE_VERSION:
        raise BundleError(f"Unsupported bundle version {version!r}, this build understands v{BUNDLE_VERSION}.")
    devices = payload.get("devices")
    if not isinstance(devices, list):
        raise BundleError("Bundle is missing a 'devices' list.")
    return [BundleEntry.from_dict(entry) for entry in devices]


def bundle_to_json(entries: list[BundleEntry]) -> str:
    """File form: pretty-printed JSON, for `--out bundle.json` and for the web
    app's file picker."""
    return json.dumps(build_bundle(entries), indent=2) + "\n"


def bundle_from_json(text: str) -> list[BundleEntry]:
    try:
        payload = json.loads(text)
    except json.JSONDecodeError as exc:
        raise BundleError(f"Not valid JSON: {exc}") from exc
    return parse_bundle(payload)


def _b64url_encode(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).decode("ascii").rstrip("=")


def _b64url_decode(text: str) -> bytes:
    padded = text + "=" * (-len(text) % 4)
    try:
        return base64.urlsafe_b64decode(padded.encode("ascii"))
    except Exception as exc:
        raise BundleError(f"Malformed base64url payload: {exc}") from exc


def _derive_key(passphrase: str, salt: bytes) -> bytes:
    kdf = PBKDF2HMAC(algorithm=hashes.SHA256(), length=_KEY_SIZE, salt=salt, iterations=_PBKDF2_ITERATIONS)
    return kdf.derive(passphrase.encode("utf-8"))


def encode_bundle(
    entries: list[BundleEntry],
    passphrase: str | None = None,
    *,
    _salt: bytes | None = None,
    _nonce: bytes | None = None,
) -> str:
    """Text form, for QR codes and URL fragments. With `passphrase`, the
    HMOLES1E form; without, the plaintext HMOLES1 form.

    `_salt`/`_nonce` exist solely so `interop.build_test_vectors()` can emit a
    reproducible encrypted vector for the TypeScript port to decrypt. Never pass
    them in real use: a reused salt/nonce pair under the same passphrase breaks
    AES-GCM outright.
    """
    raw = json.dumps(build_bundle(entries), separators=(",", ":")).encode("utf-8")
    if passphrase is None:
        return PREFIX_PLAIN + _b64url_encode(raw)

    salt = _salt if _salt is not None else secrets.token_bytes(_SALT_SIZE)
    nonce = _nonce if _nonce is not None else secrets.token_bytes(_NONCE_SIZE)
    ciphertext = AESGCM(_derive_key(passphrase, salt)).encrypt(nonce, raw, _ENCRYPTED_MAGIC)
    return PREFIX_ENCRYPTED + _b64url_encode(_ENCRYPTED_MAGIC + salt + nonce + ciphertext)


def is_encrypted_bundle(text: str) -> bool:
    """Whether `text` needs a passphrase - lets a caller (CLI, web app) prompt
    only when it actually has to."""
    return text.strip().startswith(PREFIX_ENCRYPTED)


def decode_bundle(text: str, passphrase: str | None = None) -> list[BundleEntry]:
    """Accepts every form a bundle can arrive in: the two text prefixes, a raw
    JSON file, or a URL carrying the text form in its fragment
    (`https://.../#import=HMOLES1E:...`) - the web app hands out such links, and
    users paste the whole URL rather than the payload."""
    text = text.strip()

    if "#import=" in text:
        text = text.split("#import=", 1)[1].strip()

    if text.startswith(PREFIX_ENCRYPTED):
        if passphrase is None:
            raise BundleError("This bundle is encrypted - a passphrase is required.")
        envelope = _b64url_decode(text[len(PREFIX_ENCRYPTED) :])
        offset = len(_ENCRYPTED_MAGIC)
        if len(envelope) <= offset + _SALT_SIZE + _NONCE_SIZE or not envelope.startswith(_ENCRYPTED_MAGIC):
            raise BundleError("Malformed encrypted bundle envelope.")
        salt = envelope[offset : offset + _SALT_SIZE]
        nonce = envelope[offset + _SALT_SIZE : offset + _SALT_SIZE + _NONCE_SIZE]
        ciphertext = envelope[offset + _SALT_SIZE + _NONCE_SIZE :]
        try:
            raw = AESGCM(_derive_key(passphrase, salt)).decrypt(nonce, ciphertext, _ENCRYPTED_MAGIC)
        except Exception as exc:
            raise BundleError("Could not decrypt the bundle - wrong passphrase or corrupted data.") from exc
        return bundle_from_json(raw.decode("utf-8"))

    if text.startswith(PREFIX_PLAIN):
        return bundle_from_json(_b64url_decode(text[len(PREFIX_PLAIN) :]).decode("utf-8"))

    if text.startswith("{"):
        return bundle_from_json(text)

    raise BundleError(f"Unrecognized bundle format - expected '{PREFIX_PLAIN}', '{PREFIX_ENCRYPTED}' or JSON.")

File: spa/packages/hoermoles-ble-js/src/bundle.ts
/**
 * Credential bundle - the TypeScript half of the format defined in
 * `python/packages/hoermoles-ble/src/hoermoles_ble/bundle.py`. The two files
 * implement one spec; change them together.
 *
 *     HMOLES1:<base64url(utf8 json)>            plaintext
 *     HMOLES1E:<base64url(binary envelope)>     passphrase-encrypted
 *
 * Encrypted envelope: `"HM1E" || salt(16) || nonce(12) || AES-256-GCM(ct||tag)`,
 * key via PBKDF2-HMAC-SHA256 with 600k iterations, the magic doubling as the
 * AEAD's additional data.
 *
 * `serial_no` is a **string** on the wire and stays one here. Hoermann serials
 * are uint64 (302626026414510307 on the live test device) and exceed
 * `Number.MAX_SAFE_INTEGER`, so `JSON.parse` would round the last digits away.
 * Never coerce it to a number - use it as a string, or `BigInt(...)` it.
 */

import { concatBytes, fromBase64Url, fromHex, fromUtf8, toBase64Url, toHex, utf8 } from './bytes.js';
import { aesGcmDecrypt, aesGcmEncrypt, deriveBundleKey, randomBytes } from './crypto.js';

export const BUNDLE_FORMAT = 'hoermoles-credentials';
export const BUNDLE_VERSION = 1;
export const PREFIX_PLAIN = 'HMOLES1:';
export const PREFIX_ENCRYPTED = 'HMOLES1E:';

const ENCRYPTED_MAGIC = utf8('HM1E');
const SALT_SIZE = 16;
const NONCE_SIZE = 12;

export class BundleError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'BundleError';
  }
}

/** One drive. `rootKey` is raw key material - hand it to `importRootKey()` and
 * drop the bytes as soon as possible; do not keep it in application state. */
export interface BundleEntry {
  deviceAddress: string;
  rootId: number;
  rootKey: Uint8Array;
  qrPrefix: string;
  createdUnix: number;
  /** User-chosen display name. Carried independently of the product metadata so
   * a renamed drive survives export/import, including through the CLI. */
  label?: string | null;
  productClass?: number;
  productId?: number;
  productName?: string | null;
  /** uint64 as a decimal string - see the module docstring. */
  serialNo?: string | null;
}

interface RawEntry {
  device_address: string;
  root_id: number;
  root_key_hex: string;
  qr_prefix?: string;
  created_unix?: number;
  label?: string | null;
  product_class?: number;
  product_id?: number;
  product_name?: string | null;
  /** String on the wire; a number only in bundles written before that change. */
  serial_no?: string | number | null;
}

interface RawBundle {
  format?: string;
  v?: number;
  devices?: RawEntry[];
}

function toRaw(entry: BundleEntry): RawEntry {
  const raw: RawEntry = {
    device_address: entry.deviceAddress.toUpperCase(),
    root_id: entry.rootId,
    root_key_hex: toHex(entry.rootKey),
    qr_prefix: entry.qrPrefix ?? '',
    created_unix: entry.createdUnix ?? 0,
  };
  // Only emit a label when there is one, so unnamed drives keep clean bundles.
  if (entry.label) raw.label = entry.label;
  if (entry.productClass !== undefined && entry.productId !== undefined) {
    raw.product_class = entry.productClass;
    raw.product_id = entry.productId;
    raw.product_name = entry.productName ?? null;
    raw.serial_no = entry.serialNo ?? null;
  }
  return raw;
}

/** Legacy bundles carried the serial as a JSON number; those are already
 * rounded by the time we see them, but importing something is better than
 * rejecting the whole bundle over a cosmetic field. */
function normalizeSerial(value: string | number | null | undefined): string | null {
  if (value === null || value === undefined) return null;
  return String(value);
}

function fromRaw(raw: RawEntry): BundleEntry {
  if (typeof raw?.device_address !== 'string' || typeof raw?.root_key_hex !== 'string') {
    throw new BundleError('Malformed bundle entry: missing device_address or root_key_hex');
  }
  let rootKey: Uint8Array;
  try {
    rootKey = fromHex(raw.root_key_hex);
  } catch (error) {
    throw new BundleError(`Malformed bundle entry: ${(error as Error).message}`);
  }
  const entry: BundleEntry = {
    deviceAddress: raw.device_address,
    rootId: Number(raw.root_id),
    rootKey,
    qrPrefix: raw.qr_prefix ?? '',
    createdUnix: Number(raw.created_unix ?? 0),
  };
  if (typeof raw.label === 'string' && raw.label !== '') entry.label = raw.label;
  if (raw.product_class != null && raw.product_id != null) {
    entry.productClass = Number(raw.product_class);
    entry.productId = Number(raw.product_id);
    entry.productName = raw.product_name ?? null;
    entry.serialNo = normalizeSerial(raw.serial_no);
  }
  return entry;
}

export function buildBundle(entries: readonly BundleEntry[]): RawBundle {
  return { format: BUNDLE_FORMAT, v: BUNDLE_VERSION, devices: entries.map(toRaw) };
}

export function parseBundle(payload: unknown): BundleEntry[] {
  const bundle = payload as RawBundle;
  if (typeof bundle !== 'object' || bundle === null)
    throw new BundleError('Bundle must be a JSON object.');
  if (bundle.format !== BUNDLE_FORMAT) {
    throw new BundleError(
      `Not a hoermoles credential bundle (format=${JSON.stringify(bundle.format)}).`,
    );
  }
  if (bundle.v !== BUNDLE_VERSION) {
    throw new BundleError(
      `Unsupported bundle version ${JSON.stringify(bundle.v)}, this build understands v${BUNDLE_VERSION}.`,
    );
  }
  if (!Array.isArray(bundle.devices)) throw new BundleError("Bundle is missing a 'devices' list.");
  return bundle.devices.map(fromRaw);
}

/** File form: pretty JSON, matching what the CLI's `export --out` writes. */
export function bundleToJson(entries: readonly BundleEntry[]): string {
  return JSON.stringify(buildBundle(entries), null, 2) + '\n';
}

export function bundleFromJson(text: string): BundleEntry[] {
  let payload: unknown;
  try {
    payload = JSON.parse(text);
  } catch (error) {
    throw new BundleError(`Not valid JSON: ${(error as Error).message}`);
  }
  return parseBundle(payload);
}

export function isEncryptedBundle(text: string): boolean {
  return text.trim().startsWith(PREFIX_ENCRYPTED);
}

/** Text form for QR codes and URL fragments. Passing a passphrase selects the
 * encrypted form - do that for anything leaving the device. */
export async function encodeBundle(
  entries: readonly BundleEntry[],
  passphrase?: string,
): Promise<string> {
  const raw = utf8(JSON.stringify(buildBundle(entries)));
  if (passphrase === undefined) return PREFIX_PLAIN + toBase64Url(raw);

  const salt = randomBytes(SALT_SIZE);
  const nonce = randomBytes(NONCE_SIZE);
  const key = await deriveBundleKey(passphrase, salt);
  const ciphertext = await aesGcmEncrypt(key, nonce, raw, ENCRYPTED_MAGIC);
  return PREFIX_ENCRYPTED + toBase64Url(concatBytes(ENCRYPTED_MAGIC, salt, nonce, ciphertext));
}

/** Accepts every form a bundle arrives in: both text prefixes, raw JSON, or a
 * full `https://.../#import=...` URL (users paste the whole link). */
export async function decodeBundle(text: string, passphrase?: string): Promise<BundleEntry[]> {
  let payload = text.trim();
  const fragmentIndex = payload.indexOf('#import=');
  if (fragmentIndex !== -1) payload = payload.slice(fragmentIndex + '#import='.length).trim();

  if (payload.startsWith(PREFIX_ENCRYPTED)) {
    if (passphrase === undefined)
      throw new BundleError('This bundle is encrypted - a passphrase is required.');
    let envelope: Uint8Array;
    try {
      envelope = fromBase64Url(payload.slice(PREFIX_ENCRYPTED.length));
    } catch (error) {
      throw new BundleError(`Malformed base64url payload: ${(error as Error).message}`);
    }
    const offset = ENCRYPTED_MAGIC.length;
    const magicMatches = ENCRYPTED_MAGIC.every((byte, i) => envelope[i] === byte);
    if (envelope.length <= offset + SALT_SIZE + NONCE_SIZE || !magicMatches) {
      throw new BundleError('Malformed encrypted bundle envelope.');
    }
    const salt = envelope.subarray(offset, offset + SALT_SIZE);
    const nonce = envelope.subarray(offset + SALT_SIZE, offset + SALT_SIZE + NONCE_SIZE);
    const ciphertext = envelope.subarray(offset + SALT_SIZE + NONCE_SIZE);
    let plaintext: Uint8Array;
    try {
      const key = await deriveBundleKey(passphrase, salt);
      plaintext = await aesGcmDecrypt(key, nonce, ciphertext, ENCRYPTED_MAGIC);
    } catch {
      throw new BundleError('Could not decrypt the bundle - wrong passphrase or corrupted data.');
    }
    return bundleFromJson(fromUtf8(plaintext));
  }

  if (payload.startsWith(PREFIX_PLAIN)) {
    let decoded: Uint8Array;
    try {
      decoded = fromBase64Url(payload.slice(PREFIX_PLAIN.length));
    } catch (error) {
      throw new BundleError(`Malformed base64url payload: ${(error as Error).message}`);
    }
    return bundleFromJson(fromUtf8(decoded));
  }

  if (payload.startsWith('{')) return bundleFromJson(payload);

  throw new BundleError(
    `Unrecognized bundle format - expected '${PREFIX_PLAIN}', '${PREFIX_ENCRYPTED}' or JSON.`,
  );
}
